src/main/java/engineering/swat/watch/impl/JDKDirectoryWatcher.java

```diff
@@ -0,0 +1,95 @@
+package engineering.swat.watch.impl;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.StandardWatchEventKinds;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import engineering.swat.watch.WatchEvent;
+
+public class JDKDirectoryWatcher implements Closeable {
+    private final Logger logger = LogManager.getLogger();
+    private final Path directory;
+    private final Executor exec;
+    private final Consumer<WatchEvent> eventHandler;
+    private volatile @MonotonicNonNull Closeable activeWatch;
+
+    private static final BundledSubscription<Path, List<java.nio.file.WatchEvent<?>>>
+        BUNDLED_JDK_WATCHERS = new BundledSubscription<>(JDKPoller::register);
+
+    public JDKDirectoryWatcher(Path directory, Executor exec, Consumer<WatchEvent> eventHandler) {
```

Comment:
`JDKFileWatcher` has a constructor with the same args (and does at least the same with those args as `JDKDirectoryWatcher`). Moreover, the same `safeStart`/`start` logic is present in both classes (though the code is a bit different). This overlap might be combined in a joint base class to avoid code duplication.

Reply:
Yes, but the JDKFileWatcher is actually a watch on the parent directory, that filters events for the object of interest. So they are different, although there might be a way to refactor that into some static functions in this class. Not something for this PR in my feeling, but is interesting.